### ISSUE TITLE
Types issue with useQueries

### ISSUE THREAD

**Issue by TkDodo** [tannerlinsley/react-query#3483]
useQueries type includes `Promise` in the inferred type of `.data`, which is not renderable as `React.ReactNode` with the new react 18 typings.

failing test on `beta`:

```
  it('infers returned Promises', async () => {
    const key1 = queryKey()

    // @ts-expect-error (Page component is not rendered)
    // eslint-disable-next-line
    function Page() {
      const result = useQueries({
        queries: [
          { queryKey: key1, queryFn: () => Promise.resolve(1)}
        ]
      })

      expectType<number | undefined>(result[0].data)
    }
  })
```

fails with:

```
TS2345: Argument of type 'number | Promise<number> | undefined' is not assignable to parameter of type 'number | undefined'.   Type 'Promise<number>' is not assignable to type 'number'.
```

@artysidorenko could you take a look at this please?

**Comment by artysidorenko** [comment]
Heya, had a look, added a PR to fix. Wasn't so much react-18 related; previously it had been correctly getting the `Promise<string>` => `string` inference, but got tripped up by [this union added in a recent PR](https://github.com/tannerlinsley/react-query/pull/3271/files#diff-80ebcd3dc48e85c12b2c8484c5355662ab98ceea358660eb3af536c584f6fa44R16)

**Comment by artysidorenko** [comment]
Btw, I noticed something else along the way. My understanding from reading [your other issue](https://github.com/tannerlinsley/react-query/issues/2925)  is that you want to warn with a type error anytime `queryFn` resolves to undefined?

I think at the moment (`beta` branch) it only works when it's synchronous `undefined` but not `Promise<undefined>`

Errors as expected
```typescript
// @ts-expect-error
useQuery(key, () => undefined)
```

This doesn't complain
```typescript
useQuery(key, async () => undefined)
```

Can look into it later unless that's intentional?

**Comment by TkDodo** [comment]
> Happy to look into it later unless that's intentional

oh yes good catch.  Promise<void> should also error on v4. Thank you 🙏

**Comment by TkDodo** [comment]
just to be sure, @artysidorenko, the `async () => undefined` issue wasn't fixed with #3493 right?

from what I can see:

- `useQuery(key, () => undefined)` errors, as expected ✅ 
- `useQuery(key, () => undefined)` no error ❌ 
- `useQueries({ queries: [{ queryKey: key, queryFn: () => undefined }]})` no error ❌ 
- `useQueries({ queries: [{ queryKey: key, queryFn: async () => undefined }]})` no error ❌ 

should I create a separate issue for that?

**Comment by artysidorenko** [comment]
Indeed, it wasn't fixed in #3493 , planning to look into it next. If it helps from a tracking pov, can use a separate issue, sure.

**Comment by TkDodo** [comment]
follow up created: https://github.com/tannerlinsley/react-query/issues/3516
